File: lib/tools/games/nonogram/logic/bruteforce_solver.dart
import 'dart:math';

import 'package:collection/collection.dart';
import 'package:gc_wizard/tools/games/nonogram/logic/gap_distributor.dart';
import 'package:gc_wizard/tools/games/nonogram/logic/push_solver.dart';
import 'package:gc_wizard/tools/games/nonogram/logic/util.dart';

const _cacheLimits = [2, 20];

class bruteForce extends Solver {
  Map<String, _GapResult> _solveGapCache = {};

  _GapResult? _solveGap(List<int> gap, List<int> hints) {
    var zeros = List<int>.filled(gap.length, 0);
    var ones = List<int>.filled(gap.length, 0);
    if (hints.isEmpty) {
      if (gap.contains(1)) {
        return null;
      }
      zeros.fillRange(0, zeros.length, 0);
      return _GapResult(zeros, ones);
    }
    // if (solveGap.cache == null) {
    //   solveGap.cache = [];
    // }
    if (_cacheLimits[0] <= hints.length && hints.length <= _cacheLimits[1]) {
      var key = [gap, hints].toString();
      if (_solveGapCache.containsKey(key)) {
        return _solveGapCache[key];
      }
    }
    var hint = hints[0];
    var maxIndex = gap.indexOf(1);
    if (maxIndex == -1) {
      maxIndex = gap.length;
    }
    var _hintSum = hintSum(hints);
    maxIndex = min(maxIndex, gap.length - _hintSum);
    if (maxIndex > _hintSum && !gap.contains(1)) {
      zeros.fillRange(0, zeros.length, 1);
      ones.fillRange(0, ones.length, 1);
      return _GapResult(zeros, ones);
    }
    for (var hintStart = 0; hintStart <= maxIndex; hintStart++) {
      if (hintStart + hint < gap.length && gap[hintStart + hint] == 1) {
        continue;
      }
      var rest = _solveGap(hintStart + hint + 1 < gap.length ? gap.sublist(hintStart + hint + 1) : [], hints.sublist(1));
      if (rest == null) {
        continue;
      }
      for (var k = 0; k < gap.length; k++) {
        if (k < hintStart || k == hintStart + hint) {
          zeros[k] = 1;
        } else if (k < hintStart + hint) {
          ones[k] = 1;
        } else {
          zeros[k] = ((zeros[k] != 0) || (rest.zeros[k - (hintStart + hint + 1)]) != 0 ? 1 : 0);
          ones[k]  = ((ones[k] != 0) || (rest.ones[k - (hintStart + hint + 1)]) != 0 ? 1 : 0);
        }
      }
    }
    var result = _GapResult(zeros, ones);
    _solveGapCache.addAll({[gap, hints].toString() : result});
    return result;
  }

  List<int> _solveGapWithHintList(List<int> gap, List<List<int>> hintList) {
    if (gap.contains(-1)) {
      throw const FormatException('solveGapWithHintList called with a non-gap');
    }
    var zeros = List<int>.filled(gap.length, 0);
    var ones = List<int>.filled(gap.length, 0);
    for (var hints in hintList) {
      var item = _solveGap(gap, hints);
      zeros.forEachIndexed((i, zero) => zeros[i] = (zero != 0 || ((item?.zeros[i] ?? 0) != 0) ? 1 : 0));
      ones.forEachIndexed((i, one) => ones[i] = (one != 0 || ((item?.ones[i] ?? 0) != 0) ? 1 : 0));
    }
    var result = zeros.mapIndexed((i, zero) {
      var one = ones[i];
      if (zero != 0) {
        return one != 0 ? 0 : -1;
      }
      if (one != 0) {
        return 1;
      }
      throw FormatException('Cannot fill gap ($gap, $hintList)');
    }).toList();
    // if (debugMode) {
    //   if (gap.some((x, i) => x !== result[i])) {
    //     console.log(`Gap solved: [${gap}], ${JSON.stringify(hintList)} -> ${result}`);
    //   } else {
    //     console.log(`No progress on gap: [${gap}], ${JSON.stringify(hintList)}`);
    //   }
    // }
    return result;
  }

  @override
  List<int>? solve(List<int> line, List<int> hints) {
    if (line.every((el) => el == 0)) {
      return pushSolver().solve(line, hints);
    }
    var gapResult = gapDistributor(line, hints);
    if (gapResult == null || gapResult.distributions.isEmpty) {
      print('Contradiction in line $line | $hints');
      throw FormatException('Contradiction in line $line | $hints');
    }

    var distributionsPerGap = gapResult.gaps.mapIndexed((i, gap) {
      Map<String, List<int>> set = {};
      for (var dist in gapResult.distributions) {
        var key = dist.toString();
        if (!set.containsKey(key)) {
          set.addAll({key : dist});
        }
      }
      return set.values.toList();
    }).toList();

    var result = line.sublist(0);
    Set<int> changed = {};
    gapResult.gaps.forEachIndexed((i, gap) {
    var gapResult = _solveGapWithHintList(line.sublist(gap[0], gap[1]), distributionsPerGap[i]);
      gapResult.forEachIndexed((i, item) {
        var before = result[gap[0] + i];
        if (before != item) {
          result[gap[0] + i] = item;
          changed.add(item);
        }
      });
    });
    if (changed.contains(0)) {
      throw FormatException('Contradiction in line $line | $hints');
    }

    if (changed.contains(-1)) {
      return solve(result, hints) ?? result;
    }
    return changed.contains(1) ? result : null;
  }

  @override
  bool get slowSolveSpeed {
    return true;
  }
}

class _GapResult {
  List<int> zeros = [];
  List<int> ones = [];

  _GapResult(this.zeros, this.ones);
}

File: lib/tools/games/nonogram/logic/bruteForceSolver.js
const assert = require("assert");

const util = require('../util');
const findGapDistributions = require('../gapDistributor');
const pushSolver = require('./pushSolver');

const debugMode = require('commander').debug;
const cacheLimits = [2, 20];

/**
 * @returns {{zeros: Uint8Array, ones: Uint8Array}}
 */
let solveGap = (gap, hints) => {
  let zeros = new Uint8Array(gap.length);
  let ones = new Uint8Array(gap.length);
  if (hints.length === 0) {
    if (gap.includes(1)) {
      return null;
    }
    return {
      zeros: zeros.fill(1),
      ones
    };
  }
  if (!solveGap.cache) {
    solveGap.cache = {};
  }
  if (cacheLimits[0] <= hints.length && hints.length <= cacheLimits[1]) {
    let candidate = solveGap.cache[JSON.stringify([gap, hints])];
    if (candidate) {
      return candidate;
    }
  }
  let hint = hints[0];
  let maxIndex = gap.indexOf(1);
  if (maxIndex === -1) {
    maxIndex = gap.length;
  }
  let hintSum = util.hintSum(hints);
  maxIndex = Math.min(maxIndex, gap.length - hintSum);
  if (maxIndex > hintSum && !gap.includes(1)) {
    return {
      zeros: zeros.fill(1),
      ones: ones.fill(1)
    };
  }
  for (let hintStart = 0; hintStart <= maxIndex; hintStart++) {
    if (gap[hintStart + hint] === 1) {
      continue;
    }
    let rest = solveGap(gap.slice(hintStart + hint + 1), hints.slice(1));
    if (!rest) {
      continue;
    }
    for (let k = 0; k < gap.length; k++) {
      if (k < hintStart || k === hintStart + hint) {
        zeros[k] = 1;
      } else if (k < hintStart + hint) {
        ones[k] = 1;
      } else {
        zeros[k] = zeros[k] || rest.zeros[k - (hintStart + hint + 1)];
        ones[k] = ones[k] || rest.ones[k - (hintStart + hint + 1)];
      }
    }
  }
  let result = {zeros, ones};
  solveGap.cache[JSON.stringify([gap, hints])] = result;
  return result;
};

let solveGapWithHintList = (gap, hintList) => {
  assert(!gap.includes(-1), 'solveGapWithHintList called with a non-gap');
  let zeros = new Uint8Array(gap.length);
  let ones = new Uint8Array(gap.length);
  hintList.forEach(hints => {
    let item = solveGap(gap, hints);
    zeros.forEach((zero, i) => zeros[i] = zero || item.zeros[i]);
    ones.forEach((one, i) => ones[i] = one || item.ones[i]);
  });
  let result = Array.from(zeros).map((zero, i) => {
    let one = ones[i];
    if (zero) {
      return one ? 0 : -1;
    }
    if (one) {
      return 1;
    }
    throw new Error('Cannot fill gap', gap, hintList);
  });
  if (debugMode) {
    if (gap.some((x, i) => x !== result[i])) {
      console.log(`Gap solved: [${gap}], ${JSON.stringify(hintList)} -> ${result}`);
    } else {
      console.log(`No progress on gap: [${gap}], ${JSON.stringify(hintList)}`);
    }
  }
  return result;
};

let solve = (line, hints) => {
  if (line.every(el => el === 0)) {
    return pushSolver.solve(line, hints);
  }
  let {gaps, distributions} = findGapDistributions(line, hints);
  if (debugMode) {
    console.log(`Gap distributions: ${distributions.length}`);
  }
  assert.ok(distributions.length > 0, `Contradiction in line ${line} | ${hints}`);
  let distributionsPerGap = gaps.map((gap, i) => {
    let set = new Set();
    distributions.forEach(dist => {
      set.add(JSON.stringify(dist[i]));
    });
    return Array.from(set).map(x => JSON.parse(x));
  });
  let result = line.slice();
  let changed = new Set();
  gaps.forEach((gap, i) => {
    let gapResult = solveGapWithHintList(line.slice(gap[0], gap[1]), distributionsPerGap[i]);
    gapResult.forEach((item, i) => {
      let before = result[gap[0] + i];
      if (before !== item) {
        result[gap[0] + i] = item;
        changed.add(item);
      }
    });
  });
  assert(!changed.has(0), `Contradiction in line ${line} | ${hints}`);
  if (changed.has(-1)) {
    return solve(result, hints) || result;
  }
  return changed.has(1) ? result : null;
};

solve.speed = 'slow';

module.exports = {solveGap, solveGapWithHintList, solve};
